### DISCUSSION TITLE
h.url_for not working in some templates

### DISCUSSION THREAD

**Discussion by pauldcollins** [ckan/ckan#6046] [Q&A] [answered]
**CKAN version**
2.8

**Describe the bug**
I'm having trouble using `h.url_for` in the `resource_read.html` template. I can't see what I need to import to be able to use it, so I'm not sure if it's an error.

Basically, I'm trying to copy the link from the `resource_view.html` page and use it in the `resource_read.html` template. But for some reason it breaks when I use `h.url_for`

`href="{{ h.url_for('resource_view', id=package['name'], resource_id=resource['id'], view_id=resource_view['id'], qualified=True) }}"`

**Steps to reproduce**
Copy above line into the `resource_read.html` file
https://github.com/ckan/ckan/blob/2.8/ckan/templates/package/resource_read.html

**Expected behavior**
It should go to the URL, as it does in the `resource_view.html` page.
https://github.com/ckan/ckan/blob/2.8/ckan/templates/package/snippets/resource_view.html

**Comment by Zharktas** [comment]
I doubt url_for is the issue here. When you copy and paste helper calls between templates and it fails, you probably are missing some variable from the new template. It should tell you why it fails in the logs or in the UI itself if you turn on the debug mode.

**Comment by pauldcollins** [comment]
Thanks @Zharktas for your email, you were correct - it's not the `url_for` that's the issue. I've managed to get it working, more or less - at least the page is no longer breaking but I have one last problem. 

The `view_id` is not being passed correctly, which means the **Resource View** page is producing a 404 when I click to it. The original code from the  `resource_view.html` file uses `view_id=resource_view['id']`, but **resource_view** is not available to me in this template. 

I've logged out all the available variables, but I can't see any option. To sum it up, I just want to be able to access the value of `resource_view['id']` inside the `resource_view.html` template. 

My current line of code in full is this, (but `c.id` doesn't give me the value I need): 
`href="{{ h.url_for('resource_view', id=pkg.name, resource_id=res.id, view_id=c.id, qualified=True) }}"`

Thanks for any help.

**Comment by Zharktas** [comment]
@pauldcollins in `resource_read.html` the resource view is parsed here https://github.com/ckan/ckan/blob/1e10689a4e9936069cb1c0b2fc8bf630c530c246/ckan/templates/package/resource_read.html#L114...L122 as there can be multiple resource_views.

**Comment by Zharktas** [comment]
But as this is not a bug in CKAN and is more of support request, I'll convert this to discussion.

**Comment by pauldcollins** [comment]
Hi @Zharktas thanks for moving this to a discussion. 

I've been trying to get this working the last few days. Unfortunately, the issue is I already have the `resource_view_content` block being used further down the `resource_read.html` template and I can't use it twice. 

Is there any way I can access that `resource_view['id']` variable in a different part of the `resource_read.html` template? Or do I need to create a whole new template that doesn't extend `base.html`? I'm fairly new to CKAN and not clear on how variables are passed around.

Thanks again

**Comment by anuveyatsu** [comment, marked as answer]
@pauldcollins as it was mentioned above, the `resource_views` variable should be available and it should be a list of `resource_view` dicts. For testing purposes you can try `resource_views[0]['id']` but this will always give you the first one.

**Reply by pauldcollins** [reply]
Thanks @anuveyatsu now I understand - did not realise that was available. I have logged that out. `resource_views[0]['id']` and all working.

Thanks for your help everyone.
